File: test/ModularPipelines.UnitTests/Attributes/EnumValueAttributeTests.cs
using ModularPipelines.Attributes;
using ModularPipelines.Options;
using static ModularPipelines.TestHelpers.OptionsRenderingTestHelper;

namespace ModularPipelines.UnitTests.Attributes;

public class EnumValueAttributeTests
{
    [Test]
    [Arguments(Number.One, "1")]
    [Arguments(Number.Two, "2")]
    [Arguments(Number.Three, "3")]
    public async Task Can_Parse_EnumValueAttribute(Number number, string expected)
    {
        var options = new NumberWrapper
        {
            Number = number,
        };

        var list = BuildArguments(options);
        await Assert.That(list).Contains("--number");
        await Assert.That(list).Contains(expected);
        await Assert.That(list).IsEquivalentTo(["--number", expected]);
    }

    public enum Number
    {
        [EnumValue("1")]
        One,
        [EnumValue("2")]
        Two,
        [EnumValue("3")]
        Three,
    }

    internal record NumberWrapper : CommandLineToolOptions
    {
        [CliOption("--number")]
        public Number Number { get; set; }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Attributes/EnumValueAttributeTests.fs
namespace ModularPipelines.UnitTests.FSharp.Attributes

open ModularPipelines.Attributes
open ModularPipelines.Helpers.Internal
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type Number =
    | [<EnumValue("1")>] One = 0
    | [<EnumValue("2")>] Two = 1
    | [<EnumValue("3")>] Three = 2

[<CLIMutable>]
type private NumberWrapper =
    {
        [<CliOption("--number")>]
        Number: Number
    }

type EnumValueAttributeTests() =
    member private _.ModelProvider = CommandModelProvider()
    member private _.ArgumentBuilder = CommandArgumentBuilder()
    member private this.BuildArguments(optionsObject: obj) =
        let model = this.ModelProvider.GetCommandModel(optionsObject.GetType())
        this.ArgumentBuilder.BuildArguments(model, optionsObject)

    [<Test>]
    [<Arguments(Number.One, "1")>]
    [<Arguments(Number.Two, "2")>]
    [<Arguments(Number.Three, "3")>]
    member this.Can_Parse_EnumValueAttribute(number: Number, expected: string) = async {
        let options = { Number = number }

        let list = this.BuildArguments(options)
        do! check(Assert.That(list |> Seq.contains "--number").IsTrue())
        do! check(Assert.That(list |> Seq.contains expected).IsTrue())
        do! check(Assert.That((list |> Seq.toArray) = [| "--number"; expected |]).IsTrue())
    }
